### PR TITLE
Grafana plugin

### PR DESCRIPTION
**Pull request by thoangtrvn**
1. add front-end and back-end Grana plugin as data connector

2. update ibmcloudsql:
  * thread_safe option: explicit IAM and COS session so that the library can be used by
multiple clients at the same time
  * python 3.8 compatible
  * new exceptions: catching more scenarios of service errors
  * query object files can return size other than bytes, e.g. KB, MB,
    GB, TB

3. add the pre-commit hook to ensure proper code standard
  * black is used to format code
  * pep8 and flake8 is used to check code quality

### CODE REVIEW

**Review by torsstei — CHANGES_REQUESTED** [review]

File: Python/ibmcloudsql/SQLQuery.py [hunk 1]
```diff
@@ -191,15 +210,12 @@ def wrapped(*args, **kwargs):
                 if self.project_lib is not None:
                     self.project_lib.data[sql_stmt] = {
                         "job_id": job_id,
-                        "status": status
+                        "status": status,
```
[the remark this answers is not in the document]
Comment (thoangtrvn) [reply Python/ibmcloudsql/SQLQuery.py:213]: Python supports that, this is indeed the result of 'black' code formatter that does the job. Apart from the fact that the language supports, it saves the programmers some effort when adding a new item.

**Review by torsstei — APPROVED** [review]

**Review by torsstei — APPROVED** [review]
LGTM

File: Python/ibmcloudsql/cos.py [hunk 2]
```diff
@@ -449,26 +499,38 @@ def list_cos_objects(self, cos_url):
             raise ValueError(msg)
         self.logon()
 
+        if cos_url[-1] != "/":
```
Comment (torsstei) [inline Python/ibmcloudsql/cos.py:502]: @thoangtrvn I did a system test with ibmcloudsql/test.py and discovered a regression in rename_exact_result() function, which is caused by object listing foe the result objects of a job now not returning the zero by folder object itself anymore. But rename_exact_result() is based on a certain pattern of objects being produced to decide whether it is a standard query result that can be renamed to single object or not. This is caused by this change here. Why are you appending a `/` now always? Can we undo this again?

**Review by torsstei — CHANGES_REQUESTED** [review]

**Review by torsstei — APPROVED** [review]
LGTM

File: Python/ibmcloudsql/cos.py (hunk 2 above)
Comment (thoangtrvn) [reply Python/ibmcloudsql/cos.py]: undo it.

**Pull request merged**